source/common/grpc/async_client_impl.h

```diff
@@ -292,20 +305,29 @@ class AsyncRequestImpl : public AsyncRequest,
   void onReceiveTrailingMetadata(Http::HeaderMapPtr&&) override {}
 
   void onRemoteClose(Grpc::Status::GrpcStatus status) override {
-    if (status != Grpc::Status::GrpcStatus::Ok) {
-      callbacks_.onFailure(status);
-      return;
+    Tracing::SpanFinalizerPtr finalizer = finalizer_factory_.create(request_, response_.get());
+    current_span_->setTag("grpc_status", std::to_string(status));
+
+    Grpc::Status::GrpcStatus reported_status = status;
+    if (status == Grpc::Status::GrpcStatus::Ok && response_ == nullptr) {
+      reported_status = Status::Internal;
     }
-    if (response_ == nullptr) {
-      callbacks_.onFailure(Status::Internal);
+
+    if (status != Grpc::Status::GrpcStatus::Ok) {
+      current_span_->setTag("error", "true");
+      current_span_->finishSpan(*finalizer);
+      callbacks_.onFailure(reported_status);
```

Comment:
I did think about this a bit. I was trying to push as much of the generic logic into the client as possible so it would "just work" for filters, etc. that utilized it, while offering the finalizer/factory as a mechanism for custom and context-specific additions. However, the finalizer works well for some cases, but may be an awkward interface for others, so I can't say that there's not room for more options.

Other things I considered:
- passing the span back through the existing callbacks (but this requires an interface change and is possibly just noise for folks who don't care about tracing)
- having an extra tracing-specific callback interface that can be optionally implemented and provided (a little bit heavyweight?)
- exposing the current span via accessor on the client (requires maintaining hook to the internal client, or at least to the accessor)

Since I'm not exactly sure how this is going to be used in all cases, I decided not to expand into any of the above options (yet?). However, I'm also not absolutely convinced I have the ideal interface here, so feedback and thoughts about what might work better are welcome.